### ISSUE TITLE
[Bridge RA-HPG v1.0 Implementation] Relation-aware dynamic graph block on frozen CF representation

### ISSUE THREAD

**Issue by Irthn1311** [Irthn1311/FER2013_Graph#55]
## Objective

Implement exactly one isolated **RA-HPG v1.0 (Relation-Aware Hierarchical Patch Graph)** bridge candidate to test the final pre-submission architecture-level hypothesis authorized by Issue #54 amendment `#5568328297`.

This issue is **IMPLEMENTATION ONLY**. It does not authorize FER2013 training, Kaggle scientific execution, test access, model selection, v1.1, or any other refinement.

Primary scientific question:

> Can we preserve the low-prior, convolution-free representation and regularized training behavior of accepted CF-HPG v1.2, while recovering enough fit/generalization capacity by replacing its reused Max-Relative graph blocks with a literature-grounded relation-aware **dynamic** message-passing block?

The intervention is intentionally concentrated in the **graph reasoning block**. Tokenization, patch resolution, hierarchy, classifier/readout, data augmentation, optimizer/scheduler, checkpoint policy, and test-isolation boundary remain frozen from accepted CF-HPG v1.2.

---

### Internal accepted evidence

Best generic parent is CF-HPG v1.2:

- clean-train accuracy `0.614894284022432`
- clean-train macro-F1 `0.5614369765915708`
- validation accuracy `0.5806631373641683`
- validation macro-F1 `0.5238975323290902`

Accepted intervention ladder:

- v1.0 -> v1.1 resolution: validation accuracy `+4.513792142658124 pp`
- v1.1 -> v1.2 deeper tokenizer: validation accuracy `+0.7244357759821685 pp`
- v1.2 -> v1.3 multiscale readout: validation accuracy `-0.4458066313736464 pp`

Thus spatial granularity mattered; deeper local tokenization had only a small effect; direct fine+coarse readout did not help. The remaining targeted suspect is graph/message-passing capacity.

LAP-GNN is not used as the parent because its existing seed-42 learning evidence shows a material train-validation separation: best val accuracy `0.6319308999721371`; clean-train macro-F1 `0.7562805286580438` at the best val-macro epoch vs val macro `0.601166548701511` (gap `15.511397995653287 pp`), and epoch-31 clean-train macro `0.8207655611897143` vs val macro `0.5938407974340496` (gap `22.69247637556647 pp`). The bridge therefore seeks more **relational capacity**, not a return to LAP-GNN semantic priors/readout complexity.

### External literature basis

1. **ViG — Vision GNN: An Image is Worth Graph of Nodes**, NeurIPS 2022. It establishes patch-as-node visual graph reasoning with Grapher + FFN blocks; its pyramid variants use FFN hidden expansion ratio `4`.
   - https://proceedings.neurips.cc/paper_files/paper/2022/hash/3743e69c8e47eb2e6d3afaea80e439fb-Abstract-Conference.html

2. **Face2Nodes**, Information Sciences 649 (2023) 119640, DOI `10.1016/j.ins.2023.119640`. It directly studies FER with dynamic feature-space graph construction and relation-aware graph convolution. In its FERPlus ablation, RG-Conv improved its MR-GraphConv baseline by `+1.06 pp` in training-from-scratch mode; the paper attributes the difference to relation-weighted aggregation retaining informative neighbors that max aggregation may discard. It reconstructs a graph before each RDGCN block.
   - https://doi.org/10.1016/j.ins.2023.119640

3. **DGCNN / EdgeConv**, ACM TOG 2019, DOI `10.1145/3326362`, independently supports recomputing feature-space neighborhoods layer-by-layer so affinity can evolve with learned representations.
   - https://arxiv.org/abs/1801.07829

Important boundary: Face2Nodes itself uses convolution/ResNet-based multi-scale patch embedding. **RA-HPG does not copy that component**. Only the relation-aware/dynamic graph principle is used. RA-HPG remains strict no-CNN/no-pretraining/no-anatomy.

DropMessage/DropEdge are **not** included here. They target overfit/oversmoothing; the accepted CF parent is underfit. Adding graph dropping before observing a fit/generalization shift would confound this capacity test.

---

## Exact implementation base

Use exact repository commit:

`8468dfe40ae2e6229c835b3356bb61eac1f22c77`

Create branch:

`codex/issue-55-ra-hpg-v1-relation-aware`

Create isolated candidate package:

`research/candidates/tf_ra_hpg_v1_relation_aware/`

The accepted packages must remain untouched, especially:

- `research/candidates/tf_cf_hpg/`
- `research/candidates/tf_cf_hpg_v1_1_resolution/`
- `research/candidates/tf_cf_hpg_v1_2_tokenizer/`
- `research/candidates/tf_cf_hpg_v1_3_multiscale_readout/`
- Generation-1 sources
- Step-13 sources
- `standalone/lap_gnn_tensorflow_ofix7_mid_candidate/`

---

### Input / tokenizer

- input `48x48x1` grayscale
- scale exactly `x / 127.5 - 1`
- non-overlapping `4x4` patches
- fine grid `12x12 = 144` nodes
- raw patch width `16`
- generic normalized grid-center positions in `[-1,1]^2`
- raw path: `Dense(16->96) -> GELU -> Dense(96->96)`
- position path: `Dense(2->96)`
- additive fusion after raw-local refinement
- `LayerNorm -> GELU -> Dropout(0.10)`

No convolution, CNN stem, handcrafted feature extractor, MediaPipe, landmarks, ROIs, masks, AUs, semantic slots, motif bank, query bank, transformer, or pretrained feature source.

### Hierarchy / coarse path

After exactly two fine graph blocks:

- reshape `12x12`
- exact non-overlapping `2x2` arithmetic mean
- output `6x6 = 36` coarse nodes
- parameter-free pooling
- `Dense(96->128) -> LayerNorm -> GELU`

### Readout

Exactly v1.2 coarse-only readout:

- final coarse nodes `[B,36,128]`
- global mean `[B,128]`
- global max `[B,128]`
- concat `[B,256]`
- `LayerNorm(256)`
- `Dense(256->128)`
- `GELU`
- `Dropout(0.20)`
- `Dense(128->7)` logits

Do **not** carry forward the v1.3 fine+coarse fused readout.

---

## Frozen graph topology function

Preserve the v1.2 topology rule exactly:

`E = spatial_8_neighbor_grid UNION cosine_top4(current_nodes)`

Properties:

- directed adjacency
- no self edges
- spatial 8-neighbor grid always present
- learned relation top-k = `4`
- cosine similarity
- union + dedup

Do not alter top-k, add dilation, add semantic edges, add geometry edge channels, or add anatomy priors.

### The one topology-lifecycle change

v1.2 builds one fine adjacency and reuses it across both fine blocks, and one coarse adjacency and reuses it across both coarse blocks.

RA-HPG must **recompute the same registered topology function immediately before every graph block using that block's current input nodes**:

```text
fine block 1: A_f1 = HybridGraphBuilder(current fine nodes)
fine block 2: A_f2 = HybridGraphBuilder(updated fine nodes)
coarse block 1: A_c1 = HybridGraphBuilder(current coarse nodes)
coarse block 2: A_c2 = HybridGraphBuilder(updated coarse nodes)
```

The graph builder receives the current **unnormalized block-input node tensor**, preserving v1.2 graph-builder semantics for block 1. Graph-message computation then uses the block's pre-normalized node tensor as defined below.

No graph is rebuilt inside an individual block after message passing starts.

---

## RA-HPG Relation-Aware Dynamic Graph Block

There are exactly:

- `2` fine blocks at width `D=96`
- `2` coarse blocks at width `D=128`

Each block is pre-norm residual and must implement the following registered equations.

For block input `x` and the freshly recomputed adjacency `A`:

### 1. Pre-normalization

`z_i = LayerNorm(x_i)`

### 2. Relative edge representation

For all candidate pairs:

`r_ij = z_j - z_i`

Only entries with `A_ij = true` contribute to aggregation.

### 3. Scalar relation gate

Use exactly one trainable scalar gate per edge:

`a_ij = sigmoid(Dense_D_to_1(r_ij))`

The Dense has bias and output width exactly `1`.

This is deliberately a scalar edge-importance gate, not attention softmax and not a vector gate. It keeps compute/parameter growth bounded while making neighbor contribution relation-dependent.

### 4. Degree-normalized relation-aware aggregation

`degree_i = sum_j 1[A_ij]`

`m_i = sum_j 1[A_ij] * a_ij * r_ij / max(degree_i, 1)`

Important:

- denominator is **active neighbor count**, not sum of gates
- no softmax
- no max pooling in this candidate
- no self-message
- every node must still have at least one neighbor

The degree normalization is required because the frozen hybrid graph has variable degree (grid boundary + union/dedup), unlike a fixed-k-only graph.

### 5. Graph update

Keep the v1.2 graph mixer shape:

`concat[z_i || m_i]` (`2D`)

`-> Dense(2D->D)`

`-> GELU`

`-> Dense(D->D)`

`-> Dropout(0.15)`

Residual:

`x <- x + graph_update`

### 6. FFN capacity

Use pre-norm residual FFN with literature-grounded expansion ratio `4`:

`LayerNorm(D)`

`-> Dense(D->4D)`

`-> GELU`

`-> Dense(4D->D)`

`-> Dropout(0.15)`

Residual:

`x <- x + ffn_update`

No BatchNorm, attention, extra heads, or extra graph blocks.

---

## Why expansion ratio 4 is part of this block contract

This is not a sweep. The accepted v1.2 block uses FFN expansion `2`, while both ViG's published pyramid configurations and Face2Nodes RDGCN use a `4x` node-feature transformation expansion. The bridge hypothesis is that the **graph reasoning block**, not tokenizer/readout, is the capacity bottleneck. Therefore the registered bridge block combines:

- dynamic per-block graph recomputation,
- relation-aware edge weighting instead of max-only aggregation,
- standard graph-vision FFN expansion `4`.

These three items define one preregistered **relation-capacity block intervention**. Do not vary them independently during implementation or before the first scientific result.

---

## Exact expected model identity

Accepted v1.2 identity:

- params `420839`
- trainable variables `66`
- Keras variables `76`

Registered RA-HPG arithmetic:

### FFN expansion 2 -> 4

Per block delta:

`4*D^2 + 2*D`

- two fine D=96 blocks: `74112`
- two coarse D=128 blocks: `131584`
- total FFN delta: `205696`

### Scalar relation gates

Dense `D->1` with bias:

- two fine gates: `2*(96+1) = 194`
- two coarse gates: `2*(128+1) = 258`
- total gate delta: `452`

### Total

`420839 + 205696 + 452 = 626987`

Expected exact identity:

- parameters: **`626987`**
- trainable variables: **`74`**
- Keras variables: **`84`**

If implementation of the exact registered equations does not produce this identity, STOP fail-closed and report the arithmetic/source discrepancy. Do not alter architecture to force the count.

---

## Frozen training/lifecycle configuration

Copy semantics exactly from accepted v1.2:

- dataset signature `fer2013_train28709_val3589_test3589`
- train `28709`
- validation `3589`
- seed `42`
- AdamW
- learning rate `3e-4`
- weight decay `5e-4`
- global clipnorm `1.0`
- batch `64`
- max epochs `100`
- linear warmup `5` epochs
- cosine decay to `1e-6`
- CE from logits
- label smoothing `0.05`
- checkpoint = earliest strict max `val_accuracy`
- early stopping monitor `val_loss`
- patience `15`
- min_delta `0`

Train-only image augmentation remains byte/semantic equivalent to v1.2:

- hflip p=0.5
- translation ±4 px
- rotation ±10 degrees
- contrast 0.85–1.15
- normalized brightness ±0.10
- random erasing p=0.25, area 0.02–0.10, aspect 0.5–2.0, normalized fill 0
- bilinear geometric interpolation
- REFLECT fill

Validation and clean-train evaluation have augmentation OFF.

No MixUp, CutMix, SAM, TTA, class weights, oversampling, focal loss, alternate optimizer, alternate LR, alternate seed, DropEdge, or DropMessage.

---

## Locked future comparison reference

Compare any later authorized seed-42 scientific run only to accepted CF-HPG v1.2:

- clean-train accuracy `0.614894284022432`
- clean-train macro-F1 `0.5614369765915708`
- validation accuracy `0.5806631373641683`
- validation macro-F1 `0.5238975323290902`

Future result field must be named:

`deltas_vs_cf_hpg_v1_2_pp`

Exact formulas:

- `delta_val_accuracy_pp = 100*(RA_val_acc - 0.5806631373641683)`
- `delta_val_macro_pp = 100*(RA_val_macro - 0.5238975323290902)`
- `delta_clean_train_accuracy_pp = 100*(RA_train_acc - 0.614894284022432)`
- `delta_clean_train_macro_pp = 100*(RA_train_macro - 0.5614369765915708)`

---

## Frozen future diagnostic precedence

Implementation must encode and test the following order, but **must not run FER2013 in this issue**:

1. `RA_HPG_V1_STRETCH_PASS`
2. `RA_HPG_V1_PASS`
3. `RELATION_OVERFIT_SHIFT`
4. `RELATION_STRONG_SIGNAL`
5. `RELATION_PARTIAL_SIGNAL`
6. `RELATION_FIT_WITHOUT_VAL_GAIN`
7. `RELATION_UNDERFIT_REMAINS`
8. `RELATION_INCONCLUSIVE`

### RA_HPG_V1_STRETCH_PASS
All:

- val accuracy >= `0.7000`
- val macro-F1 >= `0.6700`
- accuracy gap <= `8 pp`
- macro gap <= `8 pp`

### RA_HPG_V1_PASS
All:

- val accuracy >= `0.6500`
- val macro-F1 >= `0.6200`
- accuracy gap <= `8 pp`
- macro gap <= `8 pp`

### RELATION_OVERFIT_SHIFT

- clean-train accuracy improvement vs v1.2 >= `5.0 pp`
- AND either clean accuracy gap > `10 pp` or clean macro gap > `10 pp`

### RELATION_STRONG_SIGNAL

- validation accuracy improvement vs v1.2 >= `5.0 pp`
- AND clean-train accuracy improvement vs v1.2 >= `5.0 pp`
- unless a higher-priority label applies

### RELATION_PARTIAL_SIGNAL

- validation accuracy improvement vs v1.2 >= `3.0 pp` and < `5.0 pp`
- unless higher priority applies

### RELATION_FIT_WITHOUT_VAL_GAIN

- clean-train accuracy improvement vs v1.2 >= `5.0 pp`
- validation accuracy improvement vs v1.2 < `3.0 pp`
- both clean gaps <= `10 pp`
- unless higher priority applies

### RELATION_UNDERFIT_REMAINS

- clean-train accuracy < `0.7000`
- AND validation accuracy < `0.6200`
- unless higher priority applies

Else `RELATION_INCONCLUSIVE`.

These are practical single-seed diagnostic labels only, not statistical superiority/equivalence claims.

### Deadline kill rule

After independent review of a later authorized seed-42 run:

- `RA_HPG_V1_PASS` / `STRETCH_PASS`: freeze architecture immediately and move to confirmatory seeds + graph ablations.
- `RELATION_STRONG_SIGNAL` below PASS: retain as a serious candidate; any next action must be separately preregistered and deadline-justified.
- `RELATION_PARTIAL_SIGNAL`: at most one further intervention may be considered only if the fit/gap pattern gives a specific mechanism.
- `RELATION_FIT_WITHOUT_VAL_GAIN` or `RELATION_OVERFIT_SHIFT`: only a separately preregistered graph-regularization intervention (e.g. DropMessage) may be considered; no capacity sweep.
- `RELATION_UNDERFIT_REMAINS`: kill the bridge family for the Sep-14 submission; no further graph redesign before submission.

---

## Required implementation tests

At minimum, add focused tests proving:

1. `4x4 -> 144x16` row-major patchification unchanged from v1.2.
2. tokenizer tensor order and weights/shapes unchanged from v1.2.
3. graph topology function remains exact spatial-8 UNION cosine-top4, no self edges.
4. fine graph builder is called exactly twice per forward, once immediately before each fine block.
5. coarse graph builder is called exactly twice per forward, once immediately before each coarse block.
6. no adjacency tensor is intentionally reused across the two blocks of a scale.
7. first block graph construction receives the same pre-block node representation semantics as v1.2.
8. relation tensor orientation is exactly `r_ij = z_j - z_i`.
9. relation gate output shape is `[B,N,N,1]` and is sigmoid-bounded.
10. exact golden weighted-mean aggregation on a tiny graph. With zero gate kernel/bias, sigmoid gate=0.5 and result must equal `0.5 * mean_neighbor(z_j-z_i)`.
11. inactive/self edges contribute exactly zero.
12. graph updater is exactly `2D->D->D` with GELU.
13. FFN is exactly `D->4D->D` with GELU.
14. exactly 2 fine and 2 coarse blocks.
15. hierarchy remains exact parameter-free `12x12 -> 6x6` arithmetic mean.
16. readout remains v1.2 coarse-only mean+max, not v1.3 fused readout.
17. exact model identity `626987 / 74 / 84`.
18. no Conv2D/DepthwiseConv2D/SeparableConv2D/attention/transformer/MediaPipe/landmark/ROI/part_soft/HOG/LBP/Gabor/Laplacian semantic path.
19. `data.py` byte-identical to accepted v1.2 unless only module-docstring/import-path changes are technically unavoidable; any non-byte-identical data change requires explicit review and no scientific run authorization.
20. training configuration exactly frozen from v1.2.
21. exact v1.2 comparator and zero-delta tests.
22. all future decision precedence/boundaries including OVERFIT > STRONG/PARTIAL and FIT_WITHOUT_VAL_GAIN reachability.
23. result field name `deltas_vs_cf_hpg_v1_2_pp`.
24. CLI exposes only train CSV, val CSV, output root.
25. lexical test-path guards fail before file open.
26. `tf.function` synthetic forward/backward has finite logits/loss/all gradients and AdamW updates parameters.
27. `.keras` save/load round trip reproduces synthetic logits.
28. `git diff --check` passes.
29. v1.0/v1.1/v1.2/v1.3 candidates, Generation-1, Step-13, and frozen LAP package are unchanged.

Synthetic/tiny data only. Do not open FER train/val/test during implementation verification.

---

## Required implementation evidence

Create:

`research/evidence/tf_ra_hpg_v1_relation_aware_implementation.md`

Record:

- exact parent
- branch/head
- changed files
- model SHA-256
- graph/data/lifecycle SHA-256
- exact params/variable counts
- architecture shapes
- relation-gate equation and golden test
- dynamic graph call-count/recompute evidence
- focused tests
- combined regression tests
- tf.function/serialization checks
- test isolation
- frozen candidate diffs/checksums
- explicit statement: no FER/Kaggle/test scientific execution

Open one **Draft PR** linked to this issue and stop for research-lead review.

Final implementation status only:

`RA_HPG_V1_RELATION_AWARE_IMPLEMENTATION_ONLY`
